core/metadata_test.cc

```diff
@@ -0,0 +1,198 @@
+      dummy_meta[offset + j] = true;
+    }
+  }
+}
+
+TEST_F(MetadataTest, MultipeAttrComplexPipe) {
+  ModuleBuilder::DestroyAllModules();
+  std::vector<Module *> mods;
+  for (int i = 0; i < 10; i++) {
+    mods.push_back(create_foo());
+  }
+
+  mods[0]->AddMetadataAttr("foo", 2, bess::metadata::AccessMode::WRITE);
+  mods[1]->AddMetadataAttr("bar", 2, bess::metadata::AccessMode::WRITE);
+  mods[2]->AddMetadataAttr("foo", 2, bess::metadata::AccessMode::READ);
+  mods[2]->AddMetadataAttr("bar", 2, bess::metadata::AccessMode::READ);
+  mods[3]->AddMetadataAttr("foo", 2, bess::metadata::AccessMode::WRITE);
+  mods[4]->AddMetadataAttr("foo", 2, bess::metadata::AccessMode::READ);
+  mods[5]->AddMetadataAttr("bar", 2, bess::metadata::AccessMode::WRITE);
+  mods[6]->AddMetadataAttr("bar", 2, bess::metadata::AccessMode::READ);
+  mods[6]->AddMetadataAttr("foo", 2, bess::metadata::AccessMode::WRITE);
+  mods[7]->AddMetadataAttr("bar", 2, bess::metadata::AccessMode::WRITE);
+  mods[8]->AddMetadataAttr("foo", 2, bess::metadata::AccessMode::WRITE);
+  mods[9]->AddMetadataAttr("foo", 2, bess::metadata::AccessMode::READ);
+  mods[9]->AddMetadataAttr("bar", 2, bess::metadata::AccessMode::READ);
+
+  mods[0]->ConnectModules(0, mods[1], 0);
+  mods[1]->ConnectModules(0, mods[2], 0);
+  mods[1]->ConnectModules(1, mods[4], 0);
+  mods[0]->ConnectModules(1, mods[4], 0);
+  mods[3]->ConnectModules(0, mods[4], 0);
+  mods[4]->ConnectModules(0, mods[5], 0);
+  mods[5]->ConnectModules(0, mods[6], 0);
+  mods[7]->ConnectModules(0, mods[6], 0);
+  mods[7]->ConnectModules(1, mods[8], 0);
+  mods[8]->ConnectModules(0, mods[9], 0);
+
+  ASSERT_EQ(0, default_pipeline.ComputeMetadataOffsets());
+
+  for (const auto &it : mods) {
+    bool dummy_meta[bess::metadata::kMetadataTotalSize] = {};
+    for (size_t i = 0; i < it->num_attrs; i++) {
+      if (it->attrs[i].mode == bess::metadata::AccessMode::READ)
+        continue;
+
+      bess::metadata::mt_offset_t offset = it->attr_offsets[i];
+      for (mt_offset_t j = 0; j < it->attrs[i].size; j++) {
+        ASSERT_FALSE(dummy_meta[offset + j]);
+        dummy_meta[offset + j] = true;
+      }
+    }
+  }
+
+  ASSERT_NE(mods[0]->attr_offsets[0], mods[1]->attr_offsets[0]);
+  ASSERT_EQ(mods[0]->attr_offsets[0], mods[2]->attr_offsets[0]);
+  ASSERT_NE(mods[1]->attr_offsets[0], mods[4]->attr_offsets[0]);
+  ASSERT_EQ(mods[0]->attr_offsets[0], mods[4]->attr_offsets[0]);
+  ASSERT_EQ(mods[3]->attr_offsets[0], mods[4]->attr_offsets[0]);
+  ASSERT_EQ(mods[5]->attr_offsets[0], mods[6]->attr_offsets[0]);
+  ASSERT_NE(mods[5]->attr_offsets[0], mods[6]->attr_offsets[1]);
+  ASSERT_EQ(mods[7]->attr_offsets[0], mods[6]->attr_offsets[0]);
+  ASSERT_NE(mods[7]->attr_offsets[0], mods[6]->attr_offsets[1]);
+  ASSERT_NE(mods[7]->attr_offsets[0], mods[8]->attr_offsets[0]);
+  ASSERT_EQ(mods[7]->attr_offsets[0], mods[9]->attr_offsets[1]);
+  ASSERT_EQ(mods[8]->attr_offsets[0], mods[9]->attr_offsets[0]);
```

Comment:
@barath re: https://github.com/NetSys/bess/pull/106/files/4990a9c7e7a1bc1e6820b69ce88c76895564db86#r86116566
or did you mean that it would be sufficient to just do something like this?